docs/RFCS/cluster_upgrade_tool.md

```diff
@@ -0,0 +1,76 @@
+- Feature Name: Cluster Upgrade Tool
+- Status: draft
+- Start Date: 2016-09-15
+- Authors: Daniel Harrison
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: (one or more # from the issue tracker)
+
+# Summary
+
+A series of hooks to perform any necessary bookkeeping before a CockroachDB
+version upgrade.
+
+# Motivation
+
+Backward compatibility is crucial to a database, but can introduce considerable
+complexity to a codebase. All else being equal, it's better to keep backward
+compatibility logic siloed from the significant complexity that is inherent in
+software as complex as CockroachDB.
+
+# Detailed design
+
+Some versions of CockroachDB will require that a "system upgrade hook" be run
+before the cluster is upgraded to that version. This is expected to be
+infrequent; not every release will bump the version, only ones that need a
+migration hook.
+
+A new kv entry `SystemVersion` (ClusterVersion?) is introduced in the system
+keys to keep track of which of these hooks was the last to run. It stores a
+string version that corresponds to the CockroachDB release version requiring the
+hook. During a migration, a second kv entry is used to heartbeat and indicate
+which node is coordinating the work.
+
+A hook will migrate only from one SystemVersion to the one directly succeeding
+it. This means that a CockroachDB cluster that hasn't been upgraded for a while
+may require more than one of these hooks.
+
+When upgrading a cluster past a SystemVersion bump, the administrator is expected
+to run the `./cockroach cluster-upgrade` cli command using a binary of the
+version being upgraded to. For ease of deployment, rolling one node will also do
+this upgrade before its normal startup. Any nodes rolled after the first, but
+before the cluster is upgraded, will panic to make it very clear the roll is
```

Comment:
This seems backwards to me. It's easier for the new code to be compatible with the old data than for the migration to be performed while the old code is still running. I would prefer to roll out new code everywhere (in backwards-compatibility mode) and then perform the update once there are no more old nodes.